pkg/controller/component/lifecycle/kbagent.go

```diff
@@ -390,3 +343,44 @@ func (a *kbagent) formatError(lfa lifecycleAction, rsp proto.ActionResponse) err
 		return wrapError(err)
 	}
 }
+
+func SelectTargetPods(pods []*corev1.Pod, pod *corev1.Pod, spec *appsv1.Action) ([]*corev1.Pod, error) {
+	if spec.Exec == nil || len(spec.Exec.TargetPodSelector) == 0 {
+		return []*corev1.Pod{pod}, nil
+	}
+
+	anyPod := func() []*corev1.Pod {
+		i := rand.Int() % len(pods)
+		return []*corev1.Pod{pods[i]}
+	}
+
+	allPods := func() []*corev1.Pod {
+		return pods
+	}
+
+	podsWithRole := func() []*corev1.Pod {
+		roleName := spec.Exec.MatchingKey
+		var rolePods []*corev1.Pod
+		for i, pod := range pods {
+			if len(pod.Labels) != 0 {
+				if pod.Labels[constant.RoleLabelKey] == roleName {
+					rolePods = append(rolePods, pods[i])
+				}
+			}
+		}
```

Comment:
and if the target role pods are not available(such as seconaday), can we select a fallback pod?